### PR TITLE
Fix saving of Strings

### LINKED ISSUE #4873: Error creating new string constants
JabRef version: JabRef 5.0-dev--snapshot--2019-04-08--master--bb7b7213d
Java version: Java 1.8.0_201 (build 1.8.0_201-b09)
SO: Linux 5.0.5-1-default amd64 (openSUSE Tumbleweed 20190403)

- [X] I have tested the latest development version from http://builds.jabref.org/master/ and the problem persists

1. Select <b>Library - Edit string constants  (Ctrl+T)</b>
2. Note that the string constants are not displayed in alphabetical order in the string editor (but these should be ordered, cf. [The string editor](https://help.jabref.org/en/StringEditor))
3. Press the <b>Add new String</b> button and enter a key-value string constant.
4. Press the <b>Save</b> button.

<details>
  <summary>Exception Details</summary>
 
JabRef_20190408_205949
</details>

<details>
<summary>Log</summary>

`Uncaught exception occurred in Thread[JavaFX Application Thread,5,main]
org.jabref.model.database.KeyCollisionException: A string with that label already exists
	at org.jabref.model.database.BibDatabase.addString(BibDatabase.java:285)
	at org.jabref.model.database.BibDatabase.addStrings(BibDatabase.java:301)
	at org.jabref.gui.metadata.BibtexStringEditorDialogViewModel.save(BibtexStringEditorDialogViewModel.java:75)
	at org.jabref.gui.metadata.BibtexStringEditorDialogView.lambda$new$739(BibtexStringEditorDialogView.java:50)
	at javafx.scene.control.Dialog.impl_setResultAndClose(Dialog.java:1026)
	at javafx.scene.control.DialogPane.lambda$createButton$37(DialogPane.java:777)
	at com.sun.javafx.event.CompositeEventHandler$NormalEventHandlerRecord.handleBubblingEvent(CompositeEventHandler.java:218)
	at com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(CompositeEventHandler.java:80)
	at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:238)
	at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:191)
	at com.sun.javafx.event.CompositeEventDispatcher.dispatchBubblingEvent(CompositeEventDispatcher.java:59)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:58)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.EventUtil.fireEventImpl(EventUtil.java:74)
	at com.sun.javafx.event.EventUtil.fireEvent(EventUtil.java:49)
	at javafx.event.Event.fireEvent(Event.java:198)
	at javafx.scene.Node.fireEvent(Node.java:8411)
	at javafx.scene.control.Button.fire(Button.java:185)
	at com.sun.javafx.scene.control.behavior.ButtonBehavior.mouseReleased(ButtonBehavior.java:182)
	at com.sun.javafx.scene.control.skin.BehaviorSkinBase$1.handle(BehaviorSkinBase.java:96)
	at com.sun.javafx.scene.control.skin.BehaviorSkinBase$1.handle(BehaviorSkinBase.java:89)
	at com.sun.javafx.event.CompositeEventHandler$NormalEventHandlerRecord.handleBubblingEvent(CompositeEventHandler.java:218)
	at com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(CompositeEventHandler.java:80)
	at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:238)
	at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:191)
	at com.sun.javafx.event.CompositeEventDispatcher.dispatchBubblingEvent(CompositeEventDispatcher.java:59)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:58)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.EventUtil.fireEventImpl(EventUtil.java:74)
	at com.sun.javafx.event.EventUtil.fireEvent(EventUtil.java:49)
	at javafx.event.Event.fireEvent(Event.java:198)
	at javafx.scene.Scene$MouseHandler.process(Scene.java:3757)
	at javafx.scene.Scene$MouseHandler.access$1500(Scene.java:3485)
	at javafx.scene.Scene.impl_processMouseEvent(Scene.java:1762)
	at javafx.scene.Scene$ScenePeerListener.mouseEvent(Scene.java:2494)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler$MouseEventNotification.run(GlassViewEventHandler.java:394)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler$MouseEventNotification.run(GlassViewEventHandler.java:295)
	at java.security.AccessController.doPrivileged(Native Method)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler.lambda$handleMouseEvent$345(GlassViewEventHandler.java:432)
	at com.sun.javafx.tk.quantum.QuantumToolkit.runWithoutRenderLock(QuantumToolkit.java:389)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler.handleMouseEvent(GlassViewEventHandler.java:431)
	at com.sun.glass.ui.View.handleMouseEvent(View.java:555)
	at com.sun.glass.ui.View.notifyMouse(View.java:937)
	at com.sun.glass.ui.gtk.GtkApplication.enterNestedEventLoopImpl(Native Method)
	at com.sun.glass.ui.gtk.GtkApplication._enterNestedEventLoop(GtkApplication.java:211)
	at com.sun.glass.ui.Application.enterNestedEventLoop(Application.java:511)
	at com.sun.glass.ui.EventLoop.enter(EventLoop.java:107)
	at com.sun.javafx.tk.quantum.QuantumToolkit.enterNestedEventLoop(QuantumToolkit.java:590)
	at javafx.stage.Stage.showAndWait(Stage.java:474)
	at javafx.scene.control.HeavyweightDialog.showAndWait(HeavyweightDialog.java:162)
	at javafx.scene.control.Dialog.showAndWait(Dialog.java:341)
	at org.jabref.gui.metadata.BibtexStringEditorAction.execute(BibtexStringEditorAction.java:18)
	at org.jabref.gui.actions.JabRefAction.lambda$new$845(JabRefAction.java:29)
	at org.controlsfx.control.action.Action.handle(Action.java:419)
	at org.controlsfx.control.action.Action.handle(Action.java:64)
	at com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(CompositeEventHandler.java:86)
	at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:238)
	at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:191)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:58)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.EventUtil.fireEventImpl(EventUtil.java:74)
	at com.sun.javafx.event.EventUtil.fireEvent(EventUtil.java:49)
	at javafx.event.Event.fireEvent(Event.java:198)
	at javafx.scene.control.MenuItem.fire(MenuItem.java:462)
	at com.sun.javafx.scene.control.skin.ContextMenuContent$MenuItemContainer.doSelect(ContextMenuContent.java:1405)
	at com.sun.javafx.scene.control.skin.ContextMenuContent$MenuItemContainer.lambda$createChildren$599(ContextMenuContent.java:1358)
	at com.sun.javafx.event.CompositeEventHandler$NormalEventHandlerRecord.handleBubblingEvent(CompositeEventHandler.java:218)
	at com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(CompositeEventHandler.java:80)
	at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:238)
	at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:191)
	at com.sun.javafx.event.CompositeEventDispatcher.dispatchBubblingEvent(CompositeEventDispatcher.java:59)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:58)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.EventUtil.fireEventImpl(EventUtil.java:74)
	at com.sun.javafx.event.EventUtil.fireEvent(EventUtil.java:54)
	at javafx.event.Event.fireEvent(Event.java:198)
	at javafx.scene.Scene$MouseHandler.process(Scene.java:3757)
	at javafx.scene.Scene$MouseHandler.access$1500(Scene.java:3485)
	at javafx.scene.Scene.impl_processMouseEvent(Scene.java:1762)
	at javafx.scene.Scene$ScenePeerListener.mouseEvent(Scene.java:2494)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler$MouseEventNotification.run(GlassViewEventHandler.java:394)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler$MouseEventNotification.run(GlassViewEventHandler.java:295)
	at java.security.AccessController.doPrivileged(Native Method)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler.lambda$handleMouseEvent$345(GlassViewEventHandler.java:432)
	at com.sun.javafx.tk.quantum.QuantumToolkit.runWithoutRenderLock(QuantumToolkit.java:389)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler.handleMouseEvent(GlassViewEventHandler.java:431)
	at com.sun.glass.ui.View.handleMouseEvent(View.java:555)
	at com.sun.glass.ui.View.notifyMouse(View.java:937)
	at com.sun.glass.ui.gtk.GtkApplication._runLoop(Native Method)
	at com.sun.glass.ui.gtk.GtkApplication.lambda$null$203(GtkApplication.java:139)
	at java.lang.Thread.run(Unknown Source)
`
</details>

### PR DESCRIPTION
**Pull request by Siedlerchr**
Fix removing of Strings

Fixes #4873

The only problem I now have, I can't reliable detect/prevent dupliacte Strings,
e.g. A User can have a String with label a and content b, and a User can enter a new String a  with content c
The first String would now be overwritten.
@tobiasdiez  Do you have an idea if we can directly prevent the entering of duplicates?

The addAll method simply adds all strings and updates the content, if found. Previously it would always report a duplicate label exception.

Maybe one could check the ID value if it's present or equals.

----

- [ ] Change in CHANGELOG.md described
- [ ] Tests created for changes
- [x] Manually tested changed features in running JabRef
- [ ] Screenshots added in PR description (for bigger UI changes)
- [x] Ensured that [the git commit message is a good one](https://github.com/joelparkerhenderson/git_commit_message)
- [ ] Check documentation status (Issue created for outdated help page at [help.jabref.org](https://github.com/JabRef/help.jabref.org/issues)?)

### CODE REVIEW

File: src/main/java/org/jabref/gui/metadata/BibtexStringEditorDialogViewModel.java [hunk 1]
```diff
@@ -59,7 +59,9 @@ public void addNewString() {
     }
 
     public void removeString() {
-        allStrings.remove(selectedItemProperty.getValue());
+        BibtexStringViewModel toBeRemoved = selectedItemProperty.getValue();
+        allStrings.remove(toBeRemoved);
+        bibDatabase.removeStringByName(fromBibtexStringViewModel(toBeRemoved));
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/metadata/BibtexStringEditorDialogViewModel.java]: If the user now removes a string but presses `cancel` the string will still be removed, right?

File: src/main/java/org/jabref/model/database/BibDatabase.java [hunk 2]
```diff
@@ -293,16 +293,22 @@ public synchronized void addString(BibtexString string) throws KeyCollisionExcep
     }
 
     public void addStrings(Collection<BibtexString> stringsToAdd) {
+
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/model/database/BibDatabase.java]: I would replace this method by a `setStrings` method, that simply replaces the internal list of strings by the given one (maybe after validating that there are no duplicate keys). In addition, there should be a validation directly in the dialog.

Comment (Siedlerchr) [reply src/main/java/org/jabref/model/database/BibDatabase.java]: I implemented a validation int he dialog and replaced this by setSTrings

File: src/main/java/org/jabref/gui/metadata/BibtexStringEditorDialogView.java [hunk 3]
```diff
@@ -72,7 +77,15 @@ private void initialize() {
         new ViewModelTextFieldTableCellVisualizationFactory<BibtexStringViewModel, String>().withValidation(BibtexStringViewModel::contentValidation, visualizer).install(colContent, new DefaultStringConverter());
 
         colLabel.setOnEditCommit((CellEditEvent<BibtexStringViewModel, String> cell) -> {
-            cell.getRowValue().setLabel(cell.getNewValue());
+
+            String newLabelValue = cell.getNewValue();
+            if (cell.getTableView().getItems().stream().anyMatch(strs -> strs.labelProperty().get().equals(newLabelValue))) {
+
+                dialogService.showErrorDialogAndWait(Localization.lang("A string with the label '%0' already exists", newLabelValue));
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/metadata/BibtexStringEditorDialogView.java]: Miss fullstop at the end.

File: src/main/java/org/jabref/model/database/BibDatabase.java [hunk 4]
```diff
@@ -292,15 +292,20 @@ public synchronized void addString(BibtexString string) throws KeyCollisionExcep
...
+     * Replaces the existing lists of BibTexString with the given one
+     * No Duplicate checks are performed
+     * @param stringsToAdd The collection of strings to set
+     */
+    public void setStrings(Collection<BibtexString> stringsToAdd) {
+        Map<String, BibtexString> strs = stringsToAdd.stream().collect(Collectors.toConcurrentMap(BibtexString::getId, (bibtexStr) -> bibtexStr));
+        bibtexStrings = strs;
+    }
+
+    public void removeStringByName(BibtexString strToRemove) {
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/model/database/BibDatabase.java]: No longer required?

File: src/test/java/org/jabref/model/database/BibDatabaseTest.java [hunk 5]
```diff
@@ -126,24 +126,25 @@ public void hasStringLabelFindsString() {
     public void addStringAsCollection() {
```
Comment (tobiasdiez) [inline src/test/java/org/jabref/model/database/BibDatabaseTest.java]: please rename test methods to reflect changed method name.

**Comment by ninfito** [comment]
Thanks, now it's functional.

**Pull request merged**